aten/src/ATen/native/TensorShape.cpp

```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
   return outputs;
 }
 
+
+// Clones a tensor by cloning the underlying storage that it came from,
+// which allows us to replicate the exact strides/storage_offset in the cloned tensor.
+// Note [*_scatter ops preserve strides]
+// In order for functionalization to preserve stride correctness, the *_scatter
+// operators that it calls must preserve the striding behavior of their inputs.
+// Specifically, the output of *_scatter(base, mutated_view, ...)
+// should have identical size/stride/storage_offset to "base".
+at::Tensor clone_preserve_strides(const at::Tensor& self) {
+  TORCH_INTERNAL_ASSERT(self.has_storage());
+  // In cases where the input tensor has internal memory overlap, we cannot actually
+  // preserve the strides/storage_offset of the input tensor, because
+  // *_scatter ops will try to copy_() into the cloned tensor.
+  // However, this should **never** show up in functionalized user code;
+  // most aten ops that try to mutate a tensor with internal memory overlap would error anyway.
+  //
+  // The one place that this does come up is in autograd - if there's a select_scatter
+  // in the forward, then autograd will generate one for the backward.
+  // If the input to the select_scatter is grad_output, then this could be an expanded tensor
+  // with internal overlap.
+  if (at::has_internal_overlap(self) == at::MemOverlap::Yes) {
+    return self.clone();
+  }
+  auto dtype_size = self.dtype().itemsize();
+  auto nbytes = self.storage().sym_nbytes();
+  TORCH_INTERNAL_ASSERT(nbytes % dtype_size == 0);
```

Comment:
This is not guaranteed. You can have dtype mismatch between tensor and storage, where storage is complex, and tensor is float, and then the modulus doesn't apply.

Read meta converter for the full algorithm for how to do this.

Reply:
Actually, hmph, you're going to trace through this, maybe that doesn't work...

Reply:
Yeah, that seems pretty unfriendly to tracing.

Since we specialize on dtype, maybe I can just special case the complex/non-complex mismatch case? I'll take a look at the meta converter algorithm.

Reply:
actually - sorry - I don't think we're going to be tracing through this code?

We **want** `*_scatter` ops to show up as primitives in the trace. The only reason I need to symintify this function is because I need dynamic shape propagation through this function to work, so FakeTensor an handle `*_scatter` ops properly.

In that case, the untyped storage thing sounds right

Reply:
OK, if this is not traced then this is basically fine